corehq/project_limits/rate_counter.py

```diff
@@ -0,0 +1,201 @@
+            This is the upper limit on how long a `get` could return a stale value.
+        """
+        self.key = key
+        self.window_duration = window_duration
+        self.window_offset = window_offset
+        self.grains_per_window = grains_per_window
+        self.memoize_timeout = memoize_timeout
+        self.grain_duration = window_duration / grains_per_window
+        self.grain_counter = _FixedWindowRateCounter(
+            key=key,
+            window_duration=self.grain_duration,
+            window_offset=window_offset,
+            keep_windows=grains_per_window + 1,
+            memoize_timeout=memoize_timeout
+        )
+
+    def get(self, scope, timestamp=None):
+        if timestamp is None:
+            timestamp = time.time()
+        counts = [
+            self.grain_counter.get(scope, timestamp - i * self.grain_duration,
+                                   key_is_active=(i is 0))
+            for i in range(self.grains_per_window + 1)
+        ]
+        earliest_grain_count = counts.pop()
+        # This is the percentage of the way through the current grain we are
+        progress_in_current_grain = (timestamp % self.grain_duration) / self.grain_duration
+        # This is the count from the percentage of the earliest grain that should count
+        contribution_from_earliest = earliest_grain_count * (1 - progress_in_current_grain)
+        return sum(counts) + contribution_from_earliest
+
+    def increment(self, scope, delta=1, timestamp=None):
+        # this intentionally doesn't return because this is the active grain count,
+        # not the total that would be returned by get
+        self.grain_counter.increment(scope, delta, timestamp=timestamp)
+
+    def increment_and_get(self, scope, delta=1, timestamp=None):
+        self.increment(scope, delta, timestamp=timestamp)
+        return self.get(scope, timestamp=timestamp)
+
+
+week_rate_counter = SlidingWindowOverFixedGrainsRateCounter(
+    'week',
+    timedelta(days=7).total_seconds(),
+    grains_per_window=7,
+    memoize_timeout=timedelta(hours=6).total_seconds()
+)
+
+day_rate_counter = SlidingWindowOverFixedGrainsRateCounter(
+    'day',
+    timedelta(days=1).total_seconds(),
+    grains_per_window=4,
+    memoize_timeout=timedelta(hours=6).total_seconds()
+)
+
+hour_rate_counter = SlidingWindowOverFixedGrainsRateCounter(
+    'hour',
+    timedelta(hours=1).total_seconds(),
+    grains_per_window=3,
+    memoize_timeout=timedelta(hours=1).total_seconds()
+)
+
+minute_rate_counter = SlidingWindowOverFixedGrainsRateCounter(
+    'hour',
+    timedelta(minutes=1).total_seconds(),
+    grains_per_window=2,
+    memoize_timeout=timedelta(minutes=1).total_seconds()
+)
+
+second_rate_counter = SlidingWindowOverFixedGrainsRateCounter(
+    'hour',
+    timedelta(seconds=1).total_seconds(),
+    grains_per_window=1,
+    memoize_timeout=timedelta(seconds=1).total_seconds()
```

Comment:
Yeah the values I picked here were intentional, though I'm not sure if they're the optimal values or how much it matters exactly. `memoize_timeout` is about how long cached values held in process memory, which is faster than contacting redis but also slightly less accurate. (Writes are always correct, so it's stale-cache inaccuracy not systemic.) The weekly and daily ones I chose to cache for 6 hours since that's order of magnitude the length of an active workday, and seemed like a good balance between speed and bloating the process, but much smaller value is probably totally fine too. I'll think about your suggestion to auto-calculate them.